### PR TITLE
Add EnableShapeOfConstantFolding transformation

### PR DESCRIPTION
**Pull request by mateusztabaka**
Transpose sinking (that is used in TF frontend) disables ShapeOf constant folding which prevents some optimizations further in the pipeline. This patch introduces EnableShapeOfConstantFolding that removes DisableConstantFolding from ShapeOf nodes.

Ticket: CVS-118890

### CODE REVIEW

File: src/common/transformations/include/transformations/common_optimizations/enable_shapeof_constant_folding.hpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+//
+
+#pragma once
+
+#include <openvino/pass/graph_rewrite.hpp>
+#include <transformations_visibility.hpp>
+
+namespace ov {
+namespace pass {
+
```
Comment (evkotov) [inline src/common/transformations/include/transformations/common_optimizations/enable_shapeof_constant_folding.hpp:12]: Could you please add some comments here about transformation. What does it actually do? Maybe why we need it? Any use cases? Just in a couple of words.

File: src/common/transformations/src/transformations/common_optimizations/enable_shapeof_constant_folding.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <ngraph/pattern/op/wrap_type.hpp>
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/enable_shapeof_constant_folding.cpp:5]: please use openvino headers

**Review by evkotov — APPROVED** [review]

File: src/common/transformations/include/transformations/common_optimizations/enable_shapeof_constant_folding.hpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <openvino/pass/graph_rewrite.hpp>
```
Comment (itikhono) [inline src/common/transformations/include/transformations/common_optimizations/enable_shapeof_constant_folding.hpp:7]: minor: the same about <

File: src/common/transformations/src/transformations/transpose_sinking/ts_general.cpp [hunk 4]
```diff
@@ -77,6 +78,7 @@ bool TSGeneral::run_on_model(const std::shared_ptr<ov::Model>& f) {
         manager.register_pass<TSGeneralBackward>();
```
Comment (itikhono) [inline src/common/transformations/src/transformations/transpose_sinking/ts_general.cpp:78]: Should we disable ShapeOf here? in Backward transformations batch I mean

**Review by itikhono — APPROVED** [review]

File: src/common/transformations/src/transformations/transpose_sinking/ts_general.cpp (hunk 4 above)
Comment (mateusztabaka) [reply src/common/transformations/src/transformations/transpose_sinking/ts_general.cpp:78]: I can see that DisableShapeOfConstantFolding is registered at line 76.

Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_general.cpp:78]: I mean do we actually need it? or should we run it once in Forward transformations?

**Pull request merged**